### DISCUSSION TITLE
Decorator inheritance feels broken

### DISCUSSION THREAD

**Discussion by JesusTheHun** [storybookjs/storybook#23150] [General]
# The Problem
Global decorators are applied first.
Because of that, they are the closest to the story instead of being the furthest.

Take a look at this exemple, we define some decorators : 

```tsx
// Global decorators
export const preview: Preview = {
  decorators: [withIntlProvider, withRouter],
}

// Component decorators
export default {
  title: 'App/Login',
  render: () => <Login />,
  decorators: [withAuthProvider]
}
```

As you guessed, we wanted to add some top-level wrappers for internationalization and routing.
Let's look at what we get after setting those decorators :

```tsx
// What we get
<AuthProvider>
  <Router>
    <IntlProvider>
      <Story />
    </IntlProvider>
  </Router>
</AuthProvider>
```

Not what we expected ! Our expectations looks like that : 

```tsx
// What we expected
<IntlProvider>
  <Router>
    <AuthProvider>
      <Story />
    </AuthProvider>
  </Router>
</IntlProvider>
```

Decorators are applied in the order they are defined. 
The fact that the decorator inheritance obey to the same rule seems to break the global decorator purpose altogether.

# Solutions

## Option 1: Feature flag
Changing that behaviour would be a breaking change, but it could be done using a feature flag.

```ts
// .storybook/main.ts
const config: StorybookConfig = {
  featuresFlag: {
    reverseDecoratorsStack: true
  }
}
```

## Option 2: Accept a function 

In addition to the problem already described, there is no way to have a story that doesn't use global decorators. 
Maybe you have a few story that you don't want to wrap with them. You have to not set any global decorator, at all, just for those few stories.

If the `decorators` property accepted a function, we could have a non-breaking change, that solve both problems : 

```tsx
// Component.stories.tsx

export default {
  decorators: (globalDecorators) => [withAuthProvider, ...globalDecorators], // Custom inheritance order
}

export const NoDecorator = {
  decorators: () => [], // No decorator applied
}
```

What do you think is the best option ?

**Comment by chakAs3** [comment]
Hi @JesusTheHun, it is a really interesting topic. that we already discussed internally,  and I'm totally with the second option as it really gives you control over decorators ( order, or modification ) at the story level, it is a brilliant idea I can say.
you can even have something like.

```ts
export default {
  decorators: (globalDecorators) => {
     const firstGlobalDecorator = globalDecorators.shift()
      return [withAuthProvider, firstGlobalDecorator ], 
   }// altered decorators
}
```

**Comment by shilman** [comment]
@JesusTheHun are you sure about your post? i think decorators operate per your "expected" behavior (which is our expected behavior too):

![image](https://github.com/storybookjs/storybook/assets/488689/c61ec232-4cdb-4e8d-8642-145863baf532)

There are definitely some unsolved issues around decorator order, but I think they are more subtle than what you're claiming above.

**Comment by chakAs3** [comment]
@JesusTheHun  regarding the priority I can confirm what @shilman says, as you can see 
![image](https://github.com/storybookjs/storybook/assets/711292/a0d1fc84-a8fc-48db-8ce7-b97c050f0238)

we have been always global decorators to wrap our story in case we have UI libraries  that apply global styles

**Reply by unional** [reply]
Coming from https://github.com/storybookjs/storybook/discussions/33920,

The issue is not between `preview` vs `meta` (vs `story`).
It's the order of the `decorators` get chained/reduce together reversed the order of their execution.

i.e.: 

```tsx
{
  decorators: [fn1, fn2, fn3]
}
```

is executed as: `fn1(fn2(fn3(Story, context)))`.

It should work as "expected" if it is `fn3(fn2(fn1(Story, context)))`

**Reply by unional** [reply]
Currently I have to do something like [this](https://github.com/repobuddy/storybook/blob/main/libs/storybook/src/contexts/_story_card_scope.tsx) to reverse that order.

But it only works against local decorator array. It messed up the global order (preview > meta > story).
